enaml/src/subscription_observer.cpp

```diff
@@ -0,0 +1,332 @@
+#endif
+
+#ifdef __GNUC__
+#pragma GCC diagnostic ignored "-Wwrite-strings"
+#endif
+
+
+namespace enaml
+{
+
+// ptr to atom.api.atomref
+static cppy::ptr atomref;
+
+// POD struct - all member fields are considered private
+struct SubscriptionObserver
+{
+    PyObject_HEAD
+    PyObject* atomref;
+    PyObject* name;
+
+    static PyType_Spec TypeObject_Spec;
+
+    static PyTypeObject* TypeObject;
+
+    static bool Ready();
+
+    static bool TypeCheck( PyObject* ob );
+
+};
+
+
+namespace
+{
+
+
+PyObject*
+SubscriptionObserver_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
+{
+    PyObject* owner;
+    PyObject* name;
+    static char* kwlist[] = { "owner", "name", 0 };
+    if( !PyArg_ParseTupleAndKeywords( args, kwargs, "OU", kwlist, &owner, &name ) )
+    {
+        return 0;
+    }
+    cppy::ptr ptr( PyType_GenericNew( type, args, kwargs ) );
+    if( !ptr )
+    {
+        return 0;
+    }
+    SubscriptionObserver* self = reinterpret_cast<SubscriptionObserver*>( ptr.get() );
+
+    self->atomref = PyObject_CallOneArg(atomref.get(), owner);
+    if( !self->atomref )
+    {
+        return 0;
+    }
+    self->name =  cppy::incref( name );
+    return ptr.release();
+}
+
+
+void
+SubscriptionObserver_clear( SubscriptionObserver* self )
+{
+    Py_CLEAR( self->atomref );
+    Py_CLEAR( self->name );
+}
+
+
+int
+SubscriptionObserver_traverse( SubscriptionObserver* self, visitproc visit, void* arg )
+{
+    Py_VISIT( self->atomref );
+    Py_VISIT( self->name );
+    Py_VISIT(Py_TYPE(self));
+    return 0;
+}
+
+
+void
+SubscriptionObserver_dealloc( SubscriptionObserver* self )
+{
+    PyObject_GC_UnTrack( self );
+    SubscriptionObserver_clear( self );
+    Py_TYPE(self)->tp_free( reinterpret_cast<PyObject*>( self ) );
+}
+
+
+int
+SubscriptionObserver__bool__( SubscriptionObserver* self )
+{
+    return PyObject_IsTrue( self->atomref );
+}
+
+
+/*
+ * Calls engine update with the owner and name
+ * if self.ref:
+ *     owner = self.ref()
+ *     engine = owner._d_engine
+ *      if engine is not None:
+ *         engine.update(owner, self.name)
+ */
+PyObject*
+SubscriptionObserver_call( SubscriptionObserver* self, PyObject* args, PyObject* kwargs )
+{
+
+    if( PyObject_IsTrue( self->atomref ) )
+    {
+        cppy::ptr owner( PyObject_CallNoArgs( self->atomref ) );
+        cppy::ptr engine( owner.getattr("_d_engine") );
+        if ( !engine.is_none() )
+        {
+            cppy::ptr update_args( PyTuple_New( 2 ) );
+            if( !update_args )
+                return 0;
+            PyTuple_SET_ITEM( update_args.get(), 0, cppy::incref( owner.get() ) );
+            PyTuple_SET_ITEM( update_args.get(), 1, cppy::incref( self->name ) );
+            cppy::ptr update( engine.getattr( "update" ) );
+            return update.call( update_args );
+        }
+    }
+    Py_RETURN_NONE;
+}
+
+
+PyObject*
+SubscriptionObserver_richcompare( SubscriptionObserver* self, PyObject* other, int opid )
```

Comment:
I don't think this richcompare method is needed.

Reply:
Doesn't the atom observerpool use richcmp internally? 

Is this [new test](https://github.com/nucleic/enaml/pull/562/files#diff-70fb0bd6bd7f8242e3158b1add0f44e80ebd5ba758f11ed8dbfc422b627f27cdR33-R37)  correct?

Reply:
It does, but that feature is not needed for this `SubcriptionObserver`. A new observer is created every time the expression is evaluated, so the richcmp feature isn't necessary. It's better to short-circuit that comparison using just pointer identity.

https://github.com/nucleic/enaml/blob/main/enaml/core/standard_tracer.py#L112-L122

Reply:
We could potentially be more intelligent there and do a diff between the old and new dependencies, and only change observers as needed. But up til now, it hasn't been a bottleneck, and the current code is technically correct.